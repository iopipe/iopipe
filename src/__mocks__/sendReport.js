import _ from 'lodash';

const reports = [];

function sendReport(requestBody, config, ipAddress) {
  return new Promise(resolve => {
    const data = _.assign({}, requestBody, {
      _meta: {
        config,
        ipAddress
      }
    });
    // use a timeout to emulate some amount of network latency for the report send
    // especially useful for class.test.js - proper timeout reporting
    setTimeout(() => {
      reports.push(data);
      resolve({
        status: 200
      });
    }, 10);
  });
}

export { reports, sendReport };
