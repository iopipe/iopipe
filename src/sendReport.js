import https from 'https';

import globals from './globals';

export function sendReport(requestBody, config, ipAddress) {
  return new Promise((resolve, reject) => {
    const req = https
      .request(
        {
          hostname: ipAddress,
          servername: config.host,
          path: config.path,
          port: 443,
          method: 'POST',
          headers: { 'content-type': 'application/json' },
          agent: globals.httpsAgent,
          timeout: config.networkTimeout
        },
        res => {
          var apiResponse = '';

          res.on('data', chunk => {
            apiResponse += chunk;
          });

          res.on('end', () => {
            resolve({ status: res.statusCode, apiResponse });
          });
        }
      )
      .on('error', err => {
        reject(err);
      });

    req.write(JSON.stringify(requestBody));
    req.end();
  });
}
