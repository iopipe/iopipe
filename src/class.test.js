import _ from 'lodash';
import mockContext from 'aws-lambda-mock-context';
import isIp from 'is-ip';

import IOpipe from './index';

jest.mock('./dns');
jest.mock('./sendReport');
import * as dns from './dns';
import { reports } from './sendReport';

function createContext(opts = {}) {
  return mockContext(
    _.defaults({}, opts, {
      functionName: 'iopipe-lib-unit-tests'
    })
  );
}

function createAgent(kwargs) {
  return IOpipe(
    _.defaults(kwargs, {
      token: 'testSuite'
    })
  );
}

function fnGenerator(token, region, timeout, completedObj) {
  const iopipe = createAgent({
    token,
    url: `https://metrics-api.${region}.iopipe.com`
  });
  return iopipe(function Wrapper(event, ctx) {
    setTimeout(() => {
      completedObj[event.functionKey] = true;
      ctx.succeed(ctx.iopipe);
    }, timeout);
  });
}

function runWrappedFunction(fnToRun) {
  const ctx = createContext();
  const event = {};
  return new Promise(resolve => {
    let userFnReturnValue = undefined;
    function fnResolver(error, response) {
      return resolve({
        ctx,
        response,
        error,
        userFnReturnValue
      });
    }
    userFnReturnValue = fnToRun(event, ctx, fnResolver);
    ctx.Promise.then(success => fnResolver(null, success)).catch(fnResolver);
  });
}

function runWrappedFunctionArray(arr) {
  return Promise.all(
    arr.map(fn => {
      const ctx = mockContext();
      fn({}, ctx);
      return ctx.Promise;
    })
  );
}

// this test should run first in this file due to the nature of testing the dns promises
test('DNS promise is instantiated on library import, and reused for the coldstart invocation. New DNS promises are generated for subsequent invocations', async done => {
  try {
    const { promiseInstances } = dns;
    expect(promiseInstances.length).toBe(0);
    const iopipe = IOpipe({ token: 'testSuite' });
    expect(promiseInstances.length).toBe(1);

    const runs = [];

    const wrappedFunction = iopipe((event, ctx) => {
      runs.push(1);
      ctx.succeed('Decorate');
    });

    const run1 = await runWrappedFunction(wrappedFunction);
    expect(run1.response).toEqual('Decorate');
    expect(runs.length).toEqual(1);
    expect(promiseInstances.length).toBe(1);

    const run2 = await runWrappedFunction(wrappedFunction);
    expect(run2.response).toEqual('Decorate');
    expect(runs.length).toEqual(2);
    expect(promiseInstances.length).toBe(2);

    done();
  } catch (err) {
    console.error(err);
    throw err;
  }
});

test('Reports use different IP addresses based on config', async () => {
  const completed = {
    f1: false,
    f2: false
  };

  try {
    const wrappedFunction1 = fnGenerator(
      'ip-test-1',
      'us-west-1',
      10,
      completed
    );
    const wrappedFunction2 = fnGenerator(
      'ip-test-2',
      'us-west-2',
      5,
      completed
    );

    expect(completed.f1).toBe(false);
    expect(completed.f2).toBe(false);

    const [fn1, fn2] = await Promise.all(
      [wrappedFunction1, wrappedFunction2].map((fn, index) => {
        const ctx = mockContext();
        fn({ functionKey: `f${index + 1}` }, ctx);
        return ctx.Promise;
      })
    );
    expect(completed.f1 && completed.f2).toBe(true);
    expect(fn1.config.clientId).toBe('ip-test-1');
    expect(fn2.config.clientId).toBe('ip-test-2');
    const ips = _.chain(reports)
      .filter(r => r.client_id.match(/ip-test/))
      .map('_meta')
      .map('ipAddress')
      .value();
    expect(_.isArray(ips)).toBe(true);
    expect(ips.length).toBe(2);
    expect(_.every(ips, isIp)).toBe(true);
    expect(ips[0]).not.toEqual(ips[1]);
  } catch (err) {
    throw err;
  }
});

test('Allows ctx.iopipe.log and iopipe.log functionality', async () => {
  expect.assertions(4);
  try {
    const iopipe = createAgent({});
    const wrappedFunction = iopipe(function Wrapper(event, ctx) {
      ctx.iopipe.log('metric-1', 'foo');
      ctx.iopipe.log('metric-2', true);
      ctx.iopipe.log('metric-3', { ding: 'dong' });
      ctx.iopipe.log('metric-4', ['whoa']);
      ctx.iopipe.log('metric-5', 100);
      ctx.iopipe.log('metric-6');
      // test deprecated iopipe.log too
      iopipe.log('metric-7', true);
      ctx.succeed('all done');
    });

    const context = mockContext({ functionName: 'log-test' });
    wrappedFunction({}, context);
    const val = await context.Promise;
    expect(val).toEqual('all done');

    const metrics = _.chain(reports)
      .find(obj => obj.aws.functionName === 'log-test')
      .get('custom_metrics')
      .value();
    expect(_.isArray(metrics)).toBe(true);
    expect(metrics.length).toBe(7);
    expect(metrics).toMatchSnapshot();
  } catch (err) {
    throw err;
  }
});

test('Does not have context.iopipe.log collisions', async () => {
  try {
    const iopipe = createAgent({
      token: 'context-iopipe-log-collisions'
    });
    const wrappedFunction1 = iopipe((event, ctx) => {
      ctx.iopipe.log('func-1-log-1', true);
      ctx.iopipe.log('func-1-log-2', true);
      ctx.iopipe.log('func-1-log-3', true);
      setTimeout(() => {
        ctx.iopipe.log('func-1-log-4', true);
        ctx.succeed('wow');
      }, 6);
    });

    const wrappedFunction2 = iopipe((event, ctx) => {
      ctx.iopipe.log('func-2-log-1', true);
      ctx.iopipe.log('func-2-log-2', true);
      setTimeout(() => {
        ctx.iopipe.log('func-2-log-3', true);
      }, 2);
      setTimeout(() => {
        ctx.succeed('neat');
      }, 20);
    });

    const [fn1, fn2, fn3] = await runWrappedFunctionArray([
      wrappedFunction1,
      wrappedFunction2,
      wrappedFunction1
    ]);
    expect(fn1).toEqual('wow');
    expect(fn2).toEqual('neat');
    expect(fn3).toEqual('wow');
    const metrics = _.chain(reports)
      .filter(r => r.client_id === 'context-iopipe-log-collisions')
      .map('custom_metrics')
      .value();
    expect(_.isArray(metrics)).toBe(true);
    expect(metrics.length).toEqual(3);
    expect(metrics).toMatchSnapshot();
  } catch (err) {
    throw err;
  }
});

test('iopipe.log works, but has collisions', async () => {
  try {
    const iopipe = createAgent({
      token: 'log-collisions'
    });
    const wrappedFunction1 = iopipe((event, ctx) => {
      ctx.iopipe.log('func-1-log-1', true);
      iopipe.log('iopipe-log-func-1-log-2', true);
      ctx.iopipe.log('func-1-log-3', true);
      setTimeout(() => {
        iopipe.log('iopipe-log-func-1-log-4', true);
        ctx.succeed('wow');
      }, 6);
    });

    const wrappedFunction2 = iopipe((event, ctx) => {
      ctx.iopipe.log('func-2-log-1', true);
      ctx.iopipe.log('func-2-log-2', true);
      setTimeout(() => {
        iopipe.log('iopipe-log-func-2-log-3', true);
      }, 2);
      setTimeout(() => {
        ctx.succeed('neat');
      }, 10);
    });

    const [fn1, fn2, fn3] = await runWrappedFunctionArray([
      wrappedFunction1,
      wrappedFunction2,
      wrappedFunction1
    ]);
    expect(fn1).toEqual('wow');
    expect(fn2).toEqual('neat');
    expect(fn3).toEqual('wow');
    const metrics = _.chain(reports)
      .filter(r => r.client_id === 'log-collisions')
      .map('custom_metrics')
      .value();
    expect(_.isArray(metrics)).toBe(true);
    expect(metrics.length).toEqual(3);
    expect(metrics).toMatchSnapshot();
  } catch (err) {
    throw err;
  }
});
