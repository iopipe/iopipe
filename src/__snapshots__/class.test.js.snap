// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Allows ctx.iopipe.log and iopipe.log functionality 1`] = `
Array [
  Object {
    "n": undefined,
    "name": "metric-1",
    "s": "foo",
  },
  Object {
    "n": undefined,
    "name": "metric-2",
    "s": "true",
  },
  Object {
    "n": undefined,
    "name": "metric-3",
    "s": "{\\"ding\\":\\"dong\\"}",
  },
  Object {
    "n": undefined,
    "name": "metric-4",
    "s": "[\\"whoa\\"]",
  },
  Object {
    "n": 100,
    "name": "metric-5",
    "s": undefined,
  },
  Object {
    "n": undefined,
    "name": "metric-6",
    "s": undefined,
  },
  Object {
    "n": undefined,
    "name": "metric-7",
    "s": "true",
  },
]
`;

exports[`Does not have context.iopipe.log collisions 1`] = `
Array [
  Array [
    Object {
      "n": undefined,
      "name": "func-1-log-1",
      "s": "true",
    },
    Object {
      "n": undefined,
      "name": "func-1-log-2",
      "s": "true",
    },
    Object {
      "n": undefined,
      "name": "func-1-log-3",
      "s": "true",
    },
    Object {
      "n": undefined,
      "name": "func-1-log-4",
      "s": "true",
    },
  ],
  Array [
    Object {
      "n": undefined,
      "name": "func-1-log-1",
      "s": "true",
    },
    Object {
      "n": undefined,
      "name": "func-1-log-2",
      "s": "true",
    },
    Object {
      "n": undefined,
      "name": "func-1-log-3",
      "s": "true",
    },
    Object {
      "n": undefined,
      "name": "func-1-log-4",
      "s": "true",
    },
  ],
  Array [
    Object {
      "n": undefined,
      "name": "func-2-log-1",
      "s": "true",
    },
    Object {
      "n": undefined,
      "name": "func-2-log-2",
      "s": "true",
    },
    Object {
      "n": undefined,
      "name": "func-2-log-3",
      "s": "true",
    },
  ],
]
`;
